onset_maproom/calc.py

```diff
@@ -2,51 +2,225 @@
-    "May": 5,
-    "Jun": 6,
-    "Jul": 7,
-    "Aug": 8,
-    "Sep": 9,
-    "Oct": 10,
-    "Nov": 11,
-    "Dec": 12
-  }
-  strftimebint=strftimeb_all[strftimeb]
-  return strftimebint
+    strftimeb_all = {
+        "Jan": 1,
+        "Feb": 2,
+        "Mar": 3,
+        "Apr": 4,
+        "May": 5,
+        "Jun": 6,
+        "Jul": 7,
+        "Aug": 8,
+        "Sep": 9,
+        "Oct": 10,
+        "Nov": 11,
+        "Dec": 12,
+    }
+    strftimebint = strftimeb_all[strftimeb]
+    return strftimebint
+
+
+def daily_tobegroupedby_season(
+    daily_data, start_day, start_month, end_day, end_month, time_coord="T"
+):
+    """Returns dataset ready to be grouped by with:
+    the daily data where all days not in season of interest are dropped
+    season_starts:
+      an array where the non-dropped days are indexed by the first day of their season
+      -- to use to groupby
+    seasons_ends: an array with the dates of the end of the seasons
+    Can then apply groupby on daily_data against seasons_starts,
+    and preserving seasons_ends for the record
+    If starting day-month is 29-Feb, uses 1-Mar.
+    If ending day-month is 29-Feb, uses 1-Mar and uses < rather than <=
+    That means that the last day included in the season will be 29-Feb in leap years
+    and 28-Feb otherwise
+    """
+    # Deal with leap year cases
+    if start_day == 29 and start_month == 2:
+        start_day = 1
+        start_month = 3
+    end_day2 = end_day
+    end_month2 = end_month
+    if end_day == 29 and end_month == 2:
+        end_day2 = 1
+        end_month2 = 3
+    # Find seasons edges
+    start_edges = daily_data[time_coord].where(
+        lambda x: (x.dt.day == start_day) & (x.dt.month == start_month),
+        drop=True,
+    )
+    end_edges = daily_data[time_coord].where(
+        lambda x: (x.dt.day == end_day2) & (x.dt.month == end_month2),
+        drop=True,
+    )
+    # Drop dates outside very first and very last edges
+    #  -- this ensures we get complete seasons with regards to edges, later on
+    daily_data = daily_data.sel(**{time_coord: slice(start_edges[0], end_edges[-1])})
+    start_edges = start_edges.sel(**{time_coord: slice(start_edges[0], end_edges[-1])})
+    end_edges = end_edges.sel(
+        **{time_coord: slice(start_edges[0], end_edges[-1])}
+    ).assign_coords(**{time_coord: start_edges[time_coord]})
+    # Drops daily data not in seasons of interest
+    if end_day == 29 and end_month == 2:
+        days_in_season = (
+            daily_data[time_coord] >= start_edges.rename({time_coord: "group"})
+        ) & (daily_data[time_coord] < end_edges.rename({time_coord: "group"}))
+    else:
+        days_in_season = (
+            daily_data[time_coord] >= start_edges.rename({time_coord: "group"})
+        ) & (daily_data[time_coord] <= end_edges.rename({time_coord: "group"}))
+    days_in_season = days_in_season.sum(dim="group")
+    daily_data = daily_data.where(days_in_season == 1, drop=True)
```

Comment:
You're dropping any days that are in more than one season? How could that happen?

Reply:
it can't happen. the previous sum turns days_in_season from boolean to 1 and 0 and is used to apply OR basically. From there, days_in_season is no more boolean but 1 or 0 so need to tell it to where, needn't I?

Reply:
I see. You're not dropping days that are in more than one season, you're dropping the days that are in zero seasons.